MSI: Fix ImageResizer menu item position (dev/imageResizer)

arjunbalgovind opened the pull request:
<!-- Enter a brief description/summary of your PR here. What does it fix/what does it change/how was it tested (even manually, if necessary)? -->
## Summary of the Pull Request
This PR moves the context menu entry to the same section as the Rotate right/left options and moves the drag and drop menu entry down below the **Copy here** entry. Details of the investigations are available in issue #1110. These changes have only been verified for the MSI installer. Handling the same for MSIX will be done in a separate PR linked to the same issue.

<!-- Please review the items on the PR checklist before submitting-->
## PR Checklist
* [X] Applies to #1110 
* [X] CLA signed. If not, go over [here](https://cla.opensource.microsoft.com/microsoft/PowerToys) and sign the CLA
* [X] Tests passed

<!-- Provide a more detailed description of the PR, other things fixed or any additional comments/features here -->
## Detailed Description of the Pull Request / Additional comments
- **Details of how the context menu ordering works can be found in #1110** 
- Based on the details above, since the ShellImagePreview which contains Rotate Left/Right is registered at HKCR/SystemFileAssociation/.png/ShellEx/ContextMenuHandlers/ShellImagePreview, the same has to be done with ImageResizer for each of the supported types. Thus there are entries for bmp, dib, gif, jfif, jpe, jpeg, jpg, jxr, png, rle, tif, tiff, wdp (all the directly supported types mentioned in the ImageResizer code). Images like .ico also "work" with the ImageResizer.exe but it converts it to jpg instead, so it doesn't actually resize a .ico back to .ico.
- ShellImagePreview adds 4 entries: a separator, Rot right, Rot left and another separator. For a basic check without making the context menu handling more resource intensive, a simple check for a separator at the start index(indexMenu arg) has been added to make sure that if an older version of Windows is being used where ShellImagePreview isn't available or if it has been disabled the code will still run fine.
- The drag and drop handler index was simply increased by one because **Copy here** was the initial first value in the context menu.
<!-- Describe how you validated the behavior. Add automated tests wherever possible, but list manual validation steps taken as well -->
## Validation Steps Performed
- Build, install, run on Debug/Release x64
- Tests passed
- Checked position of drag/drop and context menu entry for .jpg, .png, .bmp.
![image](https://user-images.githubusercontent.com/32061677/72653161-81672e80-393e-11ea-8b7b-d776eb2f1c70.png)
![image](https://user-images.githubusercontent.com/32061677/72653188-98a61c00-393e-11ea-8c18-a4bd92814d49.png)

arjunbalgovind committed:
Updated vcxproj file with common project references and cpp version flags

arjunbalgovind committed:
Added module template code to dllmain and updated headers for successful build

arjunbalgovind committed:
Removed unnecessary include

arjunbalgovind committed:
Added dll to runner and add fixes to show up in PT Settings

arjunbalgovind committed:
Added settings file

arjunbalgovind committed:
Added support for enabling/disabling based on PowerRename codebase

arjunbalgovind committed:
Fixed solution file configurations

arjunbalgovind committed:
Removed Any CPU from ImageResizer csprojs

arjunbalgovind committed:
Renamed registry writing and removed build time registry addition

arjunbalgovind committed:
Added ImageResizer installation details to msi

arjunbalgovind committed:
Shifted to MII, TODO: Fix position

arjunbalgovind committed:
Incorporated ImageResizer Icon in context menu entry

arjunbalgovind committed:
Changed registry entries to SystemFileAssociations and added index changes in InsertMenuItem

arjunbalgovind committed:
Merged with dev/imageResizer

arjunbalgovind commented:
To be merged after #1113

crutkas commented:
Now for the "You're going to hate Clint" moment, does this work for MSIX?  i see you commented on #1038

arjunbalgovind commented:
@crutkas I'm not sure if it will work for MSIX and I can't verify that until I've updated the MSIX installer to include ImageResizer. That's why I created a separate issue #1110 for this, and this PR would cover the MSI part of that issue. I added the comment to #1038 and cross-referenced the issues since they are quite similar and possibly the method in this PR could help in that investigation as well. But I can't verify the MSIX part for this until I've changed the MSIX installer so I'm just getting the MSI part to work for now. I won't mark #1110 resolved until MSI and MSIX both are fixed.

arjunbalgovind committed:
Merge remote-tracking branch 'upstream/dev/imageResizer' into user/arbalgov/entryPos

arjunbalgovind committed:
Merge remote-tracking branch 'upstream/dev/imageResizer' into user/arbalgov/entryPos

arjunbalgovind commented:
@crutkas As per @yuyoyuppe's comment on #1038, a different interface altogether is used for MSIX. So this PR would apply only to MSI. The PowerRename code has both the IContextMenu and IExplorerCommand interfaces implemented and hence it supports both MSI and MSIX whereas as of now ImageResizer only has IContextMenu.

arjunbalgovind committed:
Synced with dev/imageResizer

arjunbalgovind committed:
Fixed extra newline

arjunbalgovind committed:
Fixed merge conflict

ryanbodrug-microsoft reviewed (approved):
Looks good :) Again would be great to add some tests for drag and drop vs the default case

arjunbalgovind commented:
@ryanbodrug-microsoft I'll not merge this yet so that I can get someone from the Janea team to also make sure I'm not doing anything wrong that could be problematic later for MSIX.

arjunbalgovind committed:
Merge remote-tracking branch 'upstream/dev/imageResizer' into user/arbalgov/entryPos

yuyoyuppe reviewed (approved):
LGTM. By the way, is it possible to have a separator inserted by other third-party shell extensions? What will happen if we detect it?

arjunbalgovind commented:
@yuyoyuppe Yes that is possible. In that case if a separator is present at that index then the code will move the Resize pictures option just past that separator. The reason for not doing a very comprehensive check for the Rotate Left and Rotate Right was mainly because we would anyways be shifting to MSIX and this code would not apply for that.
An other issue is that if we have to check comprehensively we would have to check the presence of all 4 entries, i.e. the separator, rot. right, rot. left and another separator, which would add computation overhead. And the other issue is that a simple string comparison of "Rotate right" will not be enough since those strings are most likely localized on Windows so we would have to add localization support for that check as well. And by fixing the structure of this ShellImagePreview as 4 menu entries, if any additional entries would be added in some future Windows update resulting in 5 entries then the code would break.
That's why I just went with a single separator check for now. If it is actually the ShellImagePreview's separator then there are no issues. If its from a third party extension then it will appear with that section which shouldn't be a major problem. Another thing is that for the separator from the third party shell extension to appear at that starting index it would have to be registered under HKCR/SystemFileAssociation/.png/ShellEx/ContextMenuHandlers/ and other img formats, so it would be likely that it is an image related extension.
